### PR TITLE
Allow developers to add custom search parameter to the engine.

### LINKED ISSUE #1382: Custom SearchParameter for local search to support offline use case
**Is your feature request related to a problem? Please describe.**
The current search implementation in the FHIR engine is based on the FHIR specification where it is specified that the default search strategy for token search as string is an exact match. We have a requirement to perform partial match to reduce the search space while displaying results on the UI. 

**Describe the solution you'd like**
Support SearchParameter for custom search strategies .

**Describe alternatives you've considered**
We have a working version with online search, but this is needed for the offline use case.

cc: @jingtang10 @fredhersch

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1382 

**Description**
1. Allows app developers to add custom search parameter to the engine via FhirEngineConfiguration.
2. Any CRUD operations on the resources after  custom search parameter is added will result in reindexing of that particular resource.
3. Engine doesn't reindex all old resources after custom search parameter is added to it. Its responsibility of the developers to reindex.

**Alternative(s) considered**
No

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
@omarismail94 does this have any interaction with your change wrt fhir versions?

**Comment by jingtang10** [comment]
@joiskash hi kashyap - please take a look at this work which might be useful to your implementation.

**Comment by joiskash** [comment]
> @joiskash hi kashyap - please take a look at this work which might be useful to your implementation.

Sure @jingtang10 ! Thanks ! 

cc: @nsabale7 @vbothe23

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt [hunk 1]
```diff
@@ -105,7 +107,17 @@ data class FhirEngineConfiguration(
...
-  val testMode: Boolean = false
+  val testMode: Boolean = false,
+  /**
+   * App developers may provide additional [SearchParameter]s to be able to search using the new
+   * parameter.
+   *
+   * NOTE: The engine doesn't reindex resources after a new [SearchParameter] is added to the
+   * engine. It is the responsibility of the app developer to reindex the resources by updating
+   * them. Any new CRUD operations on a resource after a new [SearchParameter] is added will result
+   * in the reindexing of the said resource.
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt]: done

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt [hunk 2]
```diff
@@ -105,7 +107,17 @@ data class FhirEngineConfiguration(
   val enableEncryptionIfSupported: Boolean = false,
   val databaseErrorStrategy: DatabaseErrorStrategy = UNSPECIFIED,
   val serverConfiguration: ServerConfiguration? = null,
-  val testMode: Boolean = false
+  val testMode: Boolean = false,
+  /**
+   * App developers may provide additional [SearchParameter]s to be able to search using the new
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt]: done

File: engine/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt [hunk 3]
```diff
@@ -400,11 +403,13 @@ internal object ResourceIndexer {
...
+  companion object {
+    /**
+     * The FHIR currency code system. See: https://bit.ly/30YB3ML. See:
+     * https://www.hl7.org/fhir/valueset-currencies.html.
+     */
+    private const val FHIR_CURRENCY_CODE_SYSTEM = "urn:iso:std:iso:4217"
+  }
 }
 
 data class SearchParamDefinition(val name: String, val type: SearchParamType, val path: String)
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt:415]: internal?

**Review by jingtang10 — APPROVED** [review]
Thanks Aditya this is a really good PR 👍

**Comment by jingtang10** [comment]
@stevenckngaa + @yigit + @omarismail94 in case you have any more comments.

File: engine/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt (hunk 3 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt:415]: Can't mark `SearchParamDefinition` internal as it is used in the workflow library [FhirEngineRetrieveProvider](https://github.com/google/android-fhir/blob/master/workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt#L226) as well.

**Pull request merged**